### ISSUE TITLE
Change the idiomatic name of region values from `r` to `rc`

### ISSUE THREAD

**Issue by JonathanStarup** [flix/flix#4870]
* This will remove duplicate naming for the region type and the value `reg: Region[r]`
* Smaller chance of naming collisions for three characters

**Comment by magnus-madsen** [comment]
But, on the other hand, everything else has just one letter names.

**Comment by JonathanStarup** [comment]
`ef`, `xs`, `cmp`, `len`, multiple names arguments. My comment about `r` collision stems from `Random` which would also like to be called `r`

**Comment by magnus-madsen** [comment]
I am not fully yet to pull the trigger on this. In particular, I am also considering if we need two names for regions (scope vs. region capability) and/or if we want to have region variable `r` with associated rigid variable `r!`.

**Comment by magnus-madsen** [comment]
An update on my thought process.

I think, when it comes to names, it isimportant to distinguish between different elements. That is why, for example, for the "namespace rework" I plan to have both modules and namespaces:

```
mod A.B.C {  // This is the module.
    ^^^^^    // This is *the namespace of the module*.
  ...        // This is the module.
}            // This is the module.
```

Thus when you say "namespace" you mean "A.B.C" and when you say module you mean the entire thing (which has a namespace associated with it).

Returning to the problem at hand:
```
region r { // r has type Region[r]
  let sb = StringBuilder.new(r);
  spawn (1+1) @ r
}
```

Here we have:
- A region which binds a variable `r`.
- The variable `r` is bound to a region value `r`.
- The type of the variable `r` is `Region[r']` for some 
- There are also uses of `r` inside the region.
The same name is used for many different things. How to untangle?

Internally the compiler uses `Scope` for the region construct and `Region` for the value of `r` and `Region` for the type of `r`. I don't find this much better since `Scope` and `Region` feels like it means pretty much the same thing.

An attempt could be:

```
region rh { // r has type RegionHandle[rh]
  let sb = StringBuilder.new(rh);
  spawn (1+1) @ rh
}
```

So now `rh` is a local variable of type `RegionHandle[rh]`. 

We can say that a `region` introduces a "region handle" which is a value of type `RegionHandle[...]`. It is still a bit ugly with the name of the region inside `RegionHandle`. 

Note that what I call `RegionHandle` is probably what many would call a capability. So that's also an option.

We could also go for `scope` and `ScopedHandle` to play into the buzz on structured concurrency.

In any case, I have not yet made a decision, but these are some thoughts. Let me know what you think.

**Comment by paulbutcher** [comment]
I don't have very strong opinions on most of this: I just want there to be clear a clear answer (i.e. I don't really mind what the answer is, I just want an answer to exist).

There is one thing I really do think we should change, though: `region r { ... }` should not be magic, and we should clearly differentiate between type-level and term-level variables.

Think about this from the POV of someone new to Flix. The following code is very unclear:

```
    pub def singleton(r: Region[r], k: k, v: v): MutMap[k, v, r] \ Write(r) with Order[k] =
        MutMap(ref Map.singleton(k, v) @ r)
```

As a newbie, you're thinking:

* Is there one variable called `r`? Or two?
* is the `r` in `Write` the type variable? Or is it the term variable?
* Similarly the one after the `@`

If we had different names for the two different variables, then now everything is *much* clearer.

```
    pub def singleton(reg: Region[r], k: k, v: v): MutMap[k, v, r] \ Write(r) with Order[k] =
        MutMap(ref Map.singleton(k, v) @ reg)
```

Now, I can immediately see:

* There are two different things in play, a term-level region `reg` and a type-level variable `r`
* The term-level variable is used as the parameter to the function and the argument after `@`
* The type-level variable is used within the `Write`

This whole thing is compounded by the fact that `region r { ... }` introduces both a type-level and term-level variable called `r`. I would much prefer it if they were separated somehow (again, I don't really care how, I just care that they're separated).

Perhaps `region[r] reg { ... }`?

**Comment by paulbutcher** [comment]
> Perhaps `region[r] reg { ... }`?

Incidentally, this would give us an easy way to solve the current need to sprinkle code (particularly Java interop code) with `let _ = r` statements. Instead they would become `region[r] _ { ... }`

**Comment by magnus-madsen** [comment]
Perhaps `region r { exp }` should introduce the term variable `r` with type `RegionHandle[@r]` using the special syntax `@` to show its a region?

**Comment by magnus-madsen** [comment]
I chatting with Jonathan about this next week, so by then we will have the definitive answer. But yes i think it is going in that direction.

**Comment by magnus-madsen** [comment]
As for the type level rigid variable, i think either we keep it as is or we auto generate a name that could be r plus line num plus col num.

**Comment by paulbutcher** [comment]
> As for the type level rigid variable, i think either we keep it as is or we auto generate a name that could be r plus line num plus col num.

Can I make sure I understand this properly?

You're suggesting that the idiomatic use of `region` should be `region rh { ... }`, i.e. the variable name typed by the programmer is the name of the term-level variable. There will then be a type level variable with a name of the form `r` + something to disambiguate it "magically" created as well? Do I understand that correctly?

**Comment by magnus-madsen** [comment]
Yes

**Comment by magnus-madsen** [comment]
Phone again, so excuse my brevity.

A programmer should probably almost never write the region, except it can useful for writing an explicit type annotation. If this is desired we could have *optional* syntax for naming the type variable.

**Comment by paulbutcher** [comment]
That works for me.

**Comment by magnus-madsen** [comment]
This situation is similar to type class constraints, but probably no one ever thinks about that. It's only in the top level signatures you care. But there regions are out of scope, so there is nothing to write except region variables.

**Comment by magnus-madsen** [comment]
We have agreed that region capabilities (i.e. the term/local variable) should be called `rc`.

Thus the following:

```
region rc { ... do stuff ... }
```

is now the preferred style.

For a region polymorphic function we should use `r` as the name of the type variable. Thus:

```
mod Array {
  pub def new(rc: Region[r], l: Int32): Array[a, r] \ r
}
```

@JonathanStarup @mlutze @stephentetley @paulbutcher Is the rule understandable?

**Comment by paulbutcher** [comment]
Understandable, yes 👍

It's not immediately obvious (to me, at least) what the "c" is intended to stand for? Can you help me out?

**Comment by magnus-madsen** [comment]
C is for capability following the literature.

**Comment by paulbutcher** [comment]
> C is for capability following the literature.

Oh, I see now that it was in your comment above and I missed it. Apologies.

**Comment by paulbutcher** [comment]
My feeling is that we can probably kill both the above birds with one stone? Here, for example, is code which currently needs the `let _ = r` trick:

```
    pub def toJavaList(l: List[a]): ##java.util.List = region r {
        import new java.util.ArrayList(): ##java.util.List \ Write(r) as newList;
        import java.util.List.add(##java.lang.Object): Bool \ Write(r);
        let _ = r; // Defeat unused variable error
        List.foldLeft((jl, e) -> {discard add(jl, e as ##java.lang.Object); jl}, newList(), l)
    }
```

We could change this to something like:

```
    pub def toJavaList(l: List[a]): ##java.util.List = region _ @ r {
        import new java.util.ArrayList(): ##java.util.List \ Write(r) as newList;
        import java.util.List.add(##java.lang.Object): Bool \ Write(r);
        List.foldLeft((jl, e) -> {discard add(jl, e as ##java.lang.Object); jl}, newList(), l)
    }
```

(or whatever syntax we come up with?)

**Comment by stephentetley** [comment]
I like `rc`. In Haskell `r` is quite often a name used in continuation types or the Reader monad, and I've probably brought over too much into my Flix code. `rc` is nicely distinct.

**Comment by magnus-madsen** [comment]
I think we can allow:

```
region rc: foo { ... }
```

where the programmer (optionally) gets to introduce the region type name. Subject to the usual restrictions (shadowing etc).

**Comment by magnus-madsen** [comment]
We can allow:

```
region rc: _ { ... }
```

but perhaps having the `UnusedVariable` check ignore Region variables would be better?

**Comment by magnus-madsen** [comment]
It seems there is agreement, so I could write up the steps needed?

**Comment by paulbutcher** [comment]
> I think we should allow:
> 
> ```
> region rc: foo { ... }
> ```
> 
> where the programmer (optionally) gets to introduce the region type name. Subject to the usual restrictions (shadowing etc).

Sounds good to me.

> We can allow:
> 
> ```
> region rc: _ { ... }
> ```

Minor point, but I think you meant `region _: foo { ... }`?

> but perhaps having the `UnusedVariable` check ignore Region variables would be better?

I'm not sure I see any reason why region variables should be different from others? The mechanism you proposed above gives us everything we need and fits nicely with Flix as it currently stands?

> It seems there is agreement, so I could write up the steps needed?

Works for me.

**Comment by magnus-madsen** [comment]
> Minor point, but I think you meant region _: foo { ... }?

Yes, my bad.

**Comment by magnus-madsen** [comment]
Once #5330 is merged in, I think we can do a systematic rename of `r: Region[r]` to `rc: Region[r]` and so forth.

**Comment by paulbutcher** [comment]
> It seems there is agreement, so I could write up the steps needed?

Did you get  a chance to do this @magnus-madsen? (if you did, can you point me at the issues, because I'm failing to find them)?

**Comment by magnus-madsen** [comment]
This is one:

- https://github.com/flix/flix/issues/5394

but I did not get further than that.
